Lib/ntpath.py

```diff
@@ -664,11 +664,14 @@ def _getfinalpathname_nonstrict(path):
                     raise
                 try:
                     # The OS could not resolve this path fully, so we attempt
-                    # to follow the link ourselves. If we succeed, join the tail
-                    # and return.
-                    new_path = _readlink_deep(path)
-                    if new_path != path:
+                    # to follow the link ourselves. If we succeed, keep running the loop.
+                    new_path = _readlink_deep(path, seen)
+                    new_path_normcase = normcase(new_path)
+                    if new_path_normcase in seen:
                         return join(new_path, tail) if tail else new_path
+                    else:
+                        seen.add(new_path_normcase)
+                        path = new_path
```

Comment:
Actually this is mistaken because `_readlink_deep()` always adds the path to the `seen` set. At this point, the code should `continue` if `new_path != path`. The test for a loop should be added to the root-case test on line 687, e.g. `if (path and not name) or normcase(path) in seen`. In that case, the code joins the tail and returns.

Some tests will have to be fixed that assume an incompletely resolved path. For example, in `test_realpath_broken_symlinks()`, a "broken" symlink is created that targets "missing", as well as a "broken1" symlink that targets "broken\\bar". The test assumes that "broken1" should resolve to "broken\\bar", because the current implementation doesn't process the result from `_readlink_deep()`, even though the "broken" symlink really should not be in the `realpath()` result. With this enhancement, "broken1" resolves more correctly to "missing\\bar".